### PR TITLE
752 extended override interface

### PR DESCRIPTION
**Pull request by omry**
Adding some syntax mentioned in #752 (potentially all).
The new grammar is described in [this](https://docs.google.com/document/d/1JDZGHKk4PrZHqsTTS6ao-DQOu2eVD4ULR6uAxVUR-WI/edit#) document.

Closes: #393 
Closes: #205

### Work in this diff
- [x] Choice sweep grammar
- [x] Range sweep grammar
- [x] Add range sweep support for BasicSweeper
- [x] Interval sweep grammar
- [x] Tags
- [x] Type casts
- [X] Ordering
- [x] Config group selectors (glob(*))
- [x] Documentation update
- [x] Clear new TODOs and commented out code
- [x] Add news fragments for all added features

### CODE REVIEW

File: website/docs/advanced/override_grammar/basic.md [hunk 1]
```diff
@@ -24,14 +26,14 @@ The rest are manipulating the config object.
 
 ## Grammar
 Hydra supports a rich [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) in the command line.   
-Below are the lexical rules, and [description](#Description) of the grammar.
+Blow are the parser rules from teh actual grammar.
```
Comment (jieru-hu) [inline website/docs/advanced/override_grammar/basic.md:29]: s/Blow/Below
s/teh/the

File: website/docs/advanced/override_grammar/basic.md [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
-title: Command-line syntax
+id: basic
+hide_title: true
+sidebar_label: Basic Override syntax
 ---
-You can modify your configuration via the command line. This includes:
+## Basic Override syntax
+You can manipulate your configuration with overrides (via the command line or the Compose API). This includes:
 - Modifying the the `Defaults List`
 - Modifying the config object
```
Comment (jieru-hu) [inline website/docs/advanced/override_grammar/basic.md:10]: s/matching a Config group/matching a config group

an example here would be nice

File: website/docs/advanced/override_grammar/extended.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+You can cast values, and sweeps to int, float, bool or str.
+```yaml title="Example"
+int(3.14)                  # 3 (int)
+int(value=3.14)            # 3 (int)
+float(10)                  # 10.0 (float)
+str(10)                    # "10" (str)
+bool(1)                    # true (bool)
+float(range(1,10))         # range(1.0,10.0)
+str([1,2,3])               # ['1','2','3']
+str({a:10})                # {a:'10'}
```
Comment (jieru-hu) [inline website/docs/advanced/override_grammar/extended.md:195]: str on a dict casts the value but not the key?

File: website/docs/advanced/override_grammar/basic.md [hunk 4]
```diff
@@ -142,18 +152,23 @@ foo={a:10,b:20}
...
 key={a:10, b:20},{c:30,d:40}  # and even dictionaries: ChoiceSweep([a,b], [c,d]){a,b}}
 ```
+More sweeping options are described in the [Extended Grammar page](extended).
 
 **IMPORTANT** You may need to quote your choice sweep in the shell
 
+### Functions
+Hydra supports several functions in the command line.
+See the [Extended Grammar page](extended) for more information. 
+
```
Comment (jieru-hu) [inline website/docs/advanced/override_grammar/basic.md:171]: github won't let me comment on the no change lines so putting it here

below you have `$ echo key1={a:10,b:20} key2=${HOME} key=[b]*`

but `key=[b]*` sort of format doesn't show up anywhere else (other than the grammer file), could you add a bit explanation on what it does?

File: website/docs/advanced/override_grammar/extended.md (hunk 3 above)
Comment (omry) [reply website/docs/advanced/override_grammar/extended.md:195]: correct.

File: website/docs/advanced/override_grammar/basic.md (hunk 4 above)
Comment (omry) [reply website/docs/advanced/override_grammar/basic.md:171]: You need to consult the shell documentation to understand it :)
```
omry@Coronadev:~/tmp$ touch key=bwhahwa
omry@Coronadev:~/tmp$ echo key=[b]*
key=bwhahwa
omry@Coronadev:~/tmp$ set +o braceexpand -o noglob +o histexpand
omry@Coronadev:~/tmp$ echo key=[b]*
key=[b]*
omry@Coronadev:~/tmp$ 
```

**Pull request merged**